[docs-only] Add draft of adr for spaces API. (#1827)
owncloud/ocis: docs/adr/0005-api-for-spaces.md

# 5. Graph API for Spaces

* Status: proposed
* Deciders: @butonic, @micbar, @dragotin, @hodyroff, @pmaier1
* Date: 2021-03-19

Technical Story: API to enable the concept of [Spaces](https://github.com/owncloud/enterprise/issues/3863)

## Context and Problem Statement

As one of the building blocks for Spaces in oCIS we plan to add an API that returns information about available spaces.

> Note: The term "spaces" is used here in the context of "a space where files can be saved", similar to a directory. It is not to be confused with space in the sense of file space for example.

The purpose of this new API is to give clients a very simple way to query the dynamic list of spaces, that the user has access to. Clients can provide a way better user experience with that.

> This API would even allow to provide (WebDAV-) endpoints depending on the kind and version of the client asking for it.

Reviewer: > the kind and version of the client asking for it.

What does it mean?

Reviewer 2: IMO that means the "new" clients would ask for the info, the legacy ones do not know about the api.

Author: If we wanted (or have to) we could consider in the backend code, _which_ client in _which_ version is asking and tweak the answer accordingly. I am not saying that we want to do that per default. But it could be an SOS in some situations.

This ADR discusses how this API could be designed.

## Decision Drivers

- Make it easy to work with a dynamic list of spaces of a user for the clients.
- No longer the need to make assumptions about WebDAV- and other routes in clients.
- More meta data available about spaces for a better user experience.
- Part of the bigger spaces plan.
- Important to consider in client migration scenarios, ie. in CERN.

## Considered Options

1. [Microsoft Graph API](https://developer.microsoft.com/en-us/graph) inspired API that provides the requested information.

## Decision Outcome

This the DRAFT for the API.

### API to Get Info about Spaces

ownCloud servers provide an API to query for available spaces of an user.

Most important, the API returns the WebDAV endpoint for each space. With that, clients do not have to make assumptions about WebDAV routes any more.

See [Drive item in Microsoft Graph API](https://docs.microsoft.com/de-de/graph/api/resources/onedrive?view=graph-rest-1.0)

### Get "Home folder"

> Retrieve information about the home space of a user. Note: This is only a subset of the spaces the user might have access to.

Reviewer: > This is only a subset of the spaces the user might have access to.

Multiple Spaces in `/Home` ? I thought Home is like 1 Space?

Author: Yes, `/home` is exactly one space. I will clearify the wording a bit, what I mean is: The `/me/drive` route is returning the information about the unique home for that user, but there are other spaces like shared project spaces, the shares and such.

API Call: `/me/drive`: Returns the information of the users home folder.

### Get All Spaces of a User

Retrieve a list of available spaces of a user. This includes all spaces the user has access to at that moment, also the home space.

API Call: `/me/drives`: Returns a list of spaces.

### Common Reply

The reply to both calls is either one or a list of [Drive representation objects](https://docs.microsoft.com/de-de/graph/api/resources/drive?view=graph-rest-1.0):

```
{
  "id": "string",
  "createdDateTime": "string (timestamp)",
  "description": "string",
  "driveType": "personal | business | documentLibrary",
  "eTag": "string",
  "lastModifiedDateTime": "string (timestamp)",
  "name": "string",
  "owner": { "@odata.type": "microsoft.graph.identitySet" },
  "webUrl": "url"
}

```

The meaning of the object are in ownClouds context:

1. **id** - a unique ID identifying the space
2. **driveType** - describing the type of the space.
3. **eTag** - the current ETag of the space

> 4. **owner** - an owner object to whom the space belong

Reviewer: > an owner object to whom the space belong

What about multiple owners?

Reviewer 2: good catch seems that this is limited to one user https://docs.microsoft.com/en-us/graph/api/resources/identityset?view=graph-rest-1.0

Author: Yes, this `microsoft.graph.identitySet` seems to only support one user. Strange...

Anyway, either we invent a  list of `coowner` entries, or we allow multiple owners, which I would not suggest because that would break compatibility.

Reviewer 2: @butonic you said we could use groups here, but according to spec only users, devices, application

Reviewer 3: For the CERN use case, we would need to support groups here

Author: Hm, the owner is not to be confused with who has access to the space. That has to be defined differently, and I'd say that is out of scope of *this* API. 

For project spaces, I'd say that the owner is usually the "admin god", with some coowners who can apply changes to the access rights. The idea is that spaces should not stick to a person even if he leaves...

Does that make sense?

> 5. **webUrl** - the relative Webdav path in ownCloud. It must include the `remote.php` part in oC10 installations.

Reviewer: WebDAV path can contain `$userid`, for example, the "new WebDAV endpoint".

Author: IIRC I wanted to say that it has to include the entire path underneath the domain plus installation path and not make assumptions about the remote.php/dav or webdav components of oC10.

The following driveType values are available in the first step, but might be enhanced later:

* **personal**: The users home space
* **documentLibrary**: The project spaces available for the user
* **shares**: The share jail, contains all shares for the user (*)

> 

Reviewer: Can apps extend this? For example `"driveType": Backup` introduced by some kind of "Backup" app…

Reviewer 2: according to https://docs.microsoft.com/en-us/graph/api/resources/drive?view=graph-rest-1.0  **documentLibrary** is used for sharepoint.

@dragotin Let us use onwCloud specific types.

Author: Hm, ok, but also for "personal"?

Reviewer 2: personal == home?

Reviewer 2: after talking with you, we said "stick as close as possible to the spec"

Author: @michaelstingl yes, should be extendible. How this is done is more a question for the backend service that implements the API.

@micbar I left `personal` for now because it is perfectly the same. Hope you agree.

# 5. Graph API for Spaces

* Status: proposed
* Deciders: @butonic, @micbar, @dragotin, @hodyroff, @pmaier1
* Date: 2021-03-19

Technical Story: API to enable the concept of [Spaces](https://github.com/owncloud/enterprise/issues/3863)

## Context and Problem Statement

As one of the building blocks for Spaces in oCIS we plan to add an API that returns information about available spaces.

> > Note: The term "spaces" is used here in the context of "a space where files can be saved", similar to a directory. It is not to be confused with space in the sense of file space for example.

Reviewer 4: ```suggestion
> Note: The term "spaces" is used here in the context of "a space where files can be saved", similar to a directory. It is not to be confused with space in the sense of free file space for example.
```

trying to clarify the contrast to free bytes

The purpose of this new API is to give clients a very simple way to query the dynamic list of spaces, that the user has access to. Clients can provide a way better user experience with that.

This API would even allow to provide (WebDAV-) endpoints depending on the kind and version of the client asking for it.

This ADR discusses how this API could be designed.

## Decision Drivers

- Make it easy to work with a dynamic list of spaces of a user for the clients.
- No longer the need to make assumptions about WebDAV- and other routes in clients.
- More meta data available about spaces for a better user experience.
- Part of the bigger spaces plan.
- Important to consider in client migration scenarios, ie. in CERN.

## Considered Options

1. [Microsoft Graph API](https://developer.microsoft.com/en-us/graph) inspired API that provides the requested information.

## Decision Outcome

This the DRAFT for the API.

### API to Get Info about Spaces

ownCloud servers provide an API to query for available spaces of an user.

Most important, the API returns the WebDAV endpoint for each space. With that, clients do not have to make assumptions about WebDAV routes any more.

> See [Drive item in Microsoft Graph API](https://docs.microsoft.com/de-de/graph/api/resources/onedrive?view=graph-rest-1.0)

Reviewer 4: ```suggestion
See [Drive item in Microsoft Graph API](https://docs.microsoft.com/en-us/graph/api/resources/onedrive?view=graph-rest-1.0) for an overview of `drive` and `driveItem` resources. The concrete list of drives / spaces a user has access to can be obtained on multiple endpoints.
```

### Get "Home folder"

Retrieve information about the home space of a user. Note: The user has access to more spaces. This call only returns the home space to provide API parity with the Graph API.

> API Call: `/me/drive`: Returns the information about the users home folder.

Reviewer 3: So these would be analogous to the endpoints we expose for manipulating shares `/v1.php/apps/file_sharing/api/v1/shares`? If yes, I feel there would be a disconnect between the endpoints of the OCS service and the spaces API. Maybe we can consider modifying those to adhere to the graph API specs as well? OC10 compatibility might be an issue for that, though.

Reviewer 2: IMO these are not analogous to the sharing endpoints. The sharing is confined to the "Shares" Jail, which in itself will be a drive.

I would see the drive endpoints as something new, as they are referencing "storages" not shares.

Author: For the time being both the Graph based API and the WebDAV API will coexist. Maybe we even have overlaps.

### Get All Spaces of a User

Retrieve a list of available spaces of a user. This includes all spaces the user has access to at that moment, also the home space.

> API Call: `/me/drives`: Returns a list of spaces.

Reviewer 4: ```suggestion
API Call: `/me/drives`: Returns a list of spaces.

There is also `/drives`, returning the list of spaces the user has access to. This endpoint is used to access any space by id using `/drives/{drive-id}`.
```

### Common Reply

The reply to both calls is either one or a list of [Drive representation objects](https://docs.microsoft.com/de-de/graph/api/resources/drive?view=graph-rest-1.0):

```
{
  "id": "string",
  "createdDateTime": "string (timestamp)",
  "description": "string",
  "driveType": "personal | projectSpaces | shares",
  "driveStatus": "accepted | pending | mandatory",
  "eTag": "string",
  "lastModifiedDateTime": "string (timestamp)",
  "name": "string",
  "owner": { "@odata.type": "microsoft.graph.identitySet" },
  "webUrl": "url",

>   "coowner": [ { "@odata.type": "microsoft.graph.identitySet" } ]

Reviewer 3: Maybe add a quota (bytes/files) field?

Reviewer 2: quota, yes we will

```json
{
  "deleted": 1024,
  "remaining": 1024,
  "state": "normal | nearing | critical | exceeded",
  "storagePlanInformation": {
    "upgradeAvailable": true
  },
  "total": 1024,
  "used": 1024
}
```

}

```

The meaning of the object are in ownClouds context:

1. **id** - a unique ID identifying the space

> 2. **driveType** - describing the type of the space.

Reviewer 4: ```suggestion
2. **driveType** - describing the type of the space: "personal" for users home spaces, "projectSpace" or "shares". See below.
```
I see you have customized the json above already.

Reviewer 2: commit?

Author: It is described a bit below, so I would not take it to not clutter...

> 3. **driveStatus** - telling the status

Reviewer 4: where did you get this from?

4. **eTag** - the current ETag of the space
5. **owner** - an owner object to whom the space belongs

> 6. **webUrl** - the relative Webdav path in ownCloud. It must include the `remote.php` part in oC10 installations.

Reviewer 4: `must incude` is a bit strong, because you can set up [`index.php`-less URLs](https://doc.owncloud.org/server/10.0/admin_manual/configuration/server/index_php_less_urls.html) and why relative? what about pointing to multiple domains to shard even further?

Reviewer 4: the root item contains a webDavURL property that should be used

> 7. **coowner** - optional array owner objects of the co-owners of a space (*)

Reviewer 4: ah, so I see you already added custom properties like `coowner`. Let me double check if that fits the way of extending the graph api. An example for a schema extension is in https://docs.microsoft.com/en-us/graph/extensibility-schema-groups. The open extension example at https://docs.microsoft.com/en-us/graph/extensibility-open-users jails all custom properties into an `extension` property, which is not available at the drive resource, at least not in the ms graph: https://docs.microsoft.com/en-us/graph/api/singlevaluelegacyextendedproperty-post-singlevalueextendedproperties?view=graph-rest-beta

maybe it makes sense tu use a `oc` prefix, so `coowner`  becomes `ocCoOwner` and `driveStatus` becomes `ocDriveStatus`. That way we can add properties without risking too much of a conflict?

Author: Added

The following *driveType* values are available in the first step, but might be enhanced later:

* **personal**: The users home space
* **projectSpaces**: The project spaces available for the user (*)
* **shares**: The share jail, contains all shares for the user (*)

The (*) marked types are not defined in the official MS API.

The following *driveStatus* values are available:

* **accepted**: The user has accepted the space and uses it
* **pending**: The user has not yet accepted the space , but can use it after having it accepted.
* **mandatory**: This is an mandatory space. Used for the personal- and shares-space. The user can not influence if it is visible or not, it is always available.

> 

Reviewer 4: ```suggestion
* **migrating**: The storage space is currently being migrated.
* **archived**: The storage space is archived.

```

Author: I am not sure about the migrating, which is equivalent to maintenance or just offline. All the same. However, that describes a current state of the space and might give the impression to query this API regularly to get notified about the changes. I wasn't considering that in the first place, which does not mean that it might be worse considering... 
Need to double check against the Graph API.

### Positive Consequences

- A well understood and mature API from Microsoft adopted to our needs.
- Prerequisite for Spaces in oCIS.
- Enables further steps in client development.

### Negative Consequences

- Migration impact on existing installations. Still to be investigated.

> 

Reviewer 4: ```suggestion
- Requires additional webdav endpoint that allows accessing an arbitrary storage space, either
  - with an id: `/dav/spaces/<spaceid>/relative/path/to/file.ext`, or
  - with a global path: `/dav/global/<accessible>/<mount>/<point>/relative/path/to/file.ext`, e.g. `/dav/global/projects/Golive 2021/Resources/slides.odt`

```

Reviewer 3: Yep, but for the old endpoint `/webdav/full/path`, no changes would be needed.

### Open Topics

> - Do we need a new WebDAV endpoint?

Reviewer 4: yes, see above comment.

# 5. Graph API for Spaces

* Status: proposed
* Deciders: @butonic, @micbar, @dragotin, @hodyroff, @pmaier1
* Date: 2021-03-19

Technical Story: API to enable the concept of [Spaces](https://github.com/owncloud/enterprise/issues/3863)

## Context and Problem Statement

As one of the building blocks for Spaces in oCIS we plan to add an API that returns information about available spaces. This ADR discusses the API design oriented on the Microsoft Graph API.

> Note: The term "spaces" is used here in the context of "a space where files can be saved", similar to a directory. It is not to be confused with space in the sense of free file space for example.

The purpose of this new API is to give clients a very simple way to query the dynamic list of spaces, that the user has access to. Clients can provide a better user experience with that.

> This API is supposed to be queried often, to give clients a condensed view of the available spaces for a user, but also their eTags and cTags. Hence the clients do not have to perform a PROPFIND for every space separately.

Reviewer 5: Please clarify what cTag is and how it differs from a ETAG?

Reviewer 2: CTag indicates that the content of the file has changed.
ETag indicates that content or metadata has changed.

Reviewer 6: I agree with @labkode, a section with vocabulary would help

This API would even allow to provide (WebDAV-) endpoints depending on the kind and version of the client asking for it.

## Decision Drivers

- Make it easy to work with a dynamic list of spaces of a user for the clients.
- No longer the need to make assumptions about WebDAV- and other routes in clients.
- More meta data available about spaces for a better user experience.
- Part of the bigger spaces plan.

> - Important to consider in client migration scenarios, ie. in CERN.

Reviewer 6: ```suggestion
- Important to consider in client migration scenarios.
```

## Considered Options

1. [Microsoft Graph API](https://developer.microsoft.com/en-us/graph) inspired API that provides the requested information.

## Decision Outcome

This the DRAFT for the API.

### API to Get Info about Spaces

> ownCloud servers provide an API to query for available spaces of an user.

Reviewer 6: ```suggestion
ownCloud servers provide an API to query for available spaces of a user.
```

```
{
  "id": "string",
  "createdDateTime": "string (timestamp)",
  "description": "string",
  "driveType": "personal | projectSpaces | shares",
  "oCDriveStatus": "accepted | pending | mandatory | offline",
  "lastModifiedDateTime": "string (timestamp)",
  "name": "string",
  "owner": { "@odata.type": "microsoft.graph.identitySet" },
  "quota": { "@odata.type": "microsoft.graph.quota" },
  "root":  { "@odata.type": "microsoft.graph.driveItem" },
  "webUrl": "url",
  "oCCoOwner": [ { "@odata.type": "microsoft.graph.identitySet" } ]
}

```

The meaning of the object are in ownClouds context:

1. **id** - a unique ID identifying the space
2. **driveType** - describing the type of the space.
3. **oCDriveStatus** - telling the status (*)
4. **owner** - an owner object to whom the space belongs
5. **quota** - quota information about this space
6. **root**  - the root driveItem object.
7. **webUrl** - The URL to make this space visible in the browser.
8. **oCCoOwner** - optional array owner objects of the co-owners of a space (*)

> > Note: the **root** object is a [driveItem](https://docs.microsoft.com/de-de/graph/api/resources/driveitem?view=graph-rest-1.0) and contains information about the root of the space. Especially important is the member `webDavUrl` in the root object which provides the full URL to be used by clients to address items within the space. Also it contains the fields `eTag` and `cTag`.

Reviewer 5: The `webDavUrl` property is mentioned but it doesn't appear in any of the examples.
Can you include it in the example above and also document it?

The following *driveType* values are available in the first step, but might be enhanced later:

* **personal**: The users home space
* **projectSpaces**: The project spaces available for the user (*)

> * **shares**: The share jail, contains all shares for the user (*)

Reviewer 7: This makes it sound as if all shares are mapped into this space, but the existence of `driveStatus` suggests that each share would be represented by its own space.

If so, the value should probably use the singular `share`.

Reviewer 2: Ok, this is now on the essentials about shares vs. spaces.

The SharesJail is just a collection of shares which does not include spaces (project folders)

Reviewer 6: does the name `SharesJail` imply that files there cannot be moved outside the jail?

Author: Yes, shares are *always* listed inside `shares`. That does not necessarily mean that shares is a directory. It is up to the clients to render that.

Other space types such as backup, hidden etc. can be added later as requested.

> The (*) marked types are not defined in the official MS API. They are prefixed with `oC` to avoid namespace clashes.

Reviewer 7: The marked types are not prefixed with `oC` (and I'd also prefer `oc` here).

The following *driveStatus* values are available:

> * **accepted**: The user has accepted the space and uses it

Reviewer 5: These fields represents only a perspective on how shares could work.
As a user when I'm receiving a share I may want to check the contents of it before accepting it. Accepts has collateral effect that the share will be  synchronised in the clients. 
As a user you may want to access the shares but not necessarily sync it.

* **pending**: The user has not yet accepted the space , but can use it after having it accepted.
* **mandatory**: This is an mandatory space. Used for the personal- and shares-space. The user can not influence if it is visible or not, it is always available.
* **offline**: The space is currently not available.

### Positive Consequences

- A well understood and mature API from Microsoft adopted to our needs.
- Prerequisite for Spaces in oCIS.
- Enables further steps in client development.

### Negative Consequences

- Migration impact on existing installations. Still to be investigated.
- Requires additional webdav endpoint that allows accessing an arbitrary storage space, either
  - with an id: `/dav/spaces/<spaceid>/relative/path/to/file.ext`, or
  - with a global path: `/dav/global/<accessible>/<mount>/<point>/relative/path/to/file.ext`, e.g. `/dav/global/projects/Golive 2021/Resources/slides.odt`

### Open Topics

> - What are the WebDAV pathes for Trashbin, Versions

Reviewer 6: ```suggestion
- What are the WebDAV paths for Trashbin, Versions
```
